### PR TITLE
[WIP] Overthrow gamemode

### PR DESCRIPTION
**Pull request by Carbonhell**
:cl: 
add: A new gamemode called overthrow has been added. A few syndicate mutineers spawn at roundstart, with special items such as a converting implanter and a hacked AI board, their objectives are to stage a bloodless coup. Each initial agent has its own syndicate team, and said teams can choose to cooperate or not. The team with the most points, given by the objectives, wins at roundend.
/:cl:

[why]: This is for Naksu's gamemode bounty, the guidelines are mainly his, basically the objective of those dudes is to convert/exile/kill heads and an eventual target, along with taking control of the AI. Each objective gives a certain amount of points, you get more by converting a head to your team (1.5x multiplier), if a head is converted by another team, or exiled somehow, the multiplier is 1x. If the antags are boring fucks and murderbone, the multiplier is 0.5x, to encourage conversion and exiling without necessarily avoiding killing. Each head has a base point value defined in the overthrow objective dm with defines which i will paste here too:
#define CAPPTS		5
#define AIPTS		5
#define HOPPTS		4
#define HOSPTS		4
#define TARGETPTS	4
#define CEPTS		3
#define CMOPTS		3
#define RDPTS		3
#define WARDENPTS	2
#define SECPTS		1

multipliers:
#define CONVERTED_OURS	1.5
#define CONVERTED		1
#define EXILED			1
#define KILLED			0.5

there are defines for warden and sec too, i am not totally sure whether to make converting them give points too or not, but right now it's not coded in.
Leaders can convert others (aka wake them, the "lore" is about sleeping syndicate agents") which will "reveal" their antag status, along with giving them a storage implant with a random item (mostly syndicate items that cost < 5 TCs, hand picked into a static list of the antagonist datum)
Round ends once shuttle reaches centcom, points are calculated and the team with the most points is declared winner. The crew has to fight /all/ teams to win, obviously. Teams can also fight each other to get more points, sabotage others to make them get less points, and such.
I did some very basic tests(basically checking if the gamemode itself runs without runtimes), it sorta works now but it needs futher fixes.

### CODE REVIEW

**Comment by ExcessiveUseOfCobblestone** [comment]
It's a better rev without updating rev smh

**Comment by Tlaltecuhtli** [comment]
rev but you kill ppl with ai?

**Comment by XDTM** [comment]
Wait, so every single crewmember is a sleeper agent?

**Comment by Carbonhell** [comment]
That's just the fluff behind it, to explain them magically obtaining a storage implant with an item in it, but it's just like rev conversion @XDTM

**Comment by XDTM** [comment]
Seems like there's an incentive for each team to kill the opposing team's convertees, at that will negate the point disparity (0.5 for both instead of 1.5 for one and 1.0 for the other), unless i misunderstoon point attribution.

**Comment by XDTM** [comment]
I'd go as far as to suggest an OOC directive for the teams to keep the crew alive (self-defense aside), because i feel like the playerbase really has an issue of "antag=can murderbone" in most cases

**Comment by Carbonhell** [comment]
Technically yes, that's where the team is supposed to do some teamwork and defend the targets they convert instead of going a la rev aka convert and forget
also i'd rather fix any issue codewise since i'm mainly a hippie coder and i want this to work on all forks aswell without rules getting in

**Comment by ExcessiveUseOfCobblestone** [comment]
How do you lose?

**Comment by Mickyan** [comment]
Interesting idea, I have mixed feelings about giving everyone random traitor items, and the "lore" behind it is rather flimsy, on a minor note. 

What if instead each leader had an uplink that generated TC through conversions/gaining points (may start empty or with only a small amount of TC). Gives the leader more power to set a clear strategy and distribute equipment.

There's the added benefit that people can no longer metagame the round type with 100% certainty just because they find a traitor item laying around

**Comment by Carbonhell** [comment]
the "just because they find a traitor item laying around" isn't really metagame-able, or more than it is currently.  Also, points can change in the round, if you convert a head the modifier is 1.5x but another team can kill said head to lower it to 0.5x, so i assume you mean something like checking every x minutes the current amount of points and generating y telecrystals based on said amount, which could be done but could also make those antags really powerful and give them too much traitor gear, when simply having a head on your side helps your team a lot

**Comment by Carbonhell** [comment]
Only one team can win at round end, aka the one with the most points. You lose if your team gets fully killed,or another team gathers more points than you

**Comment by fludd12** [comment]
I dunno, I find the whole Paranoia tabletop rpg style, everyones-a-sleeper-agent-of someone-thing to be pretty funny. Might need some more development to work out, but I like it

**Comment by Mickyan** [comment]
I didn't mean to say this PR would cause metagaming the round

I think periodically checking for points to award tc could make for some interesting tactics, do you rush your objectives and risk getting found out to get some good equipment early on or play it safe without the help of traitor items?
It's just a matter of balancing the TC intake to be sensibly low. Give them a max cap if necessary.

**Comment by Iamgoofball** [comment]
>>>conversion gamemodes

**Comment by ExcessiveUseOfCobblestone** [comment]
If only one team can win why would they help each other?

If someone is converted and I can't convert them again OR the switch conversion process is too long i'm just going to off them. I'll very readily take the 50% vs risking the other team letting them escape and get 150% percent.

One thing I can't wait to see is the Policy Discussion, or even better the first ahelp, regarding "is killing a person worth 150 for any reason detrimental to the team?"

**Comment by octareenroon91** [comment]
Is this likely to play out like gangs?

**Comment by Nabski89** [comment]
The big changes from original gangs is that there's not a dominator/hard end the round object, and there's an attempt to discourage mass murder by a point scheme.

Giving someone recurring points based off thing makes it very gangy where if you don't have your boss alive you are fucked, or even worse robustin gangs where everyone just is constantly shitting out guns and knives.

**Comment by ExcessiveUseOfCobblestone** [comment]
If it doesn't play like gang it will play like sling where the important people hide in [location] while the minions go out and kill everyone else.

**Comment by octareenroon91** [comment]
>discourage mass murder by a point scheme

Would it work even if a live neutral player was worth more than a dead foe?

File: code/game/gamemodes/overthrow/overthrow.dm [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+	var/sleeping_agents = required_enemies + round(num_players()*0.1)
+
+	for (var/i in 1 to sleeping_agents)
+		if (!antag_candidates.len)
+			break
+		var/datum/mind/sleeping_agent = antag_pick(antag_candidates)
+		antag_candidates -= sleeping_agent
+		initial_agents += sleeping_agent
+		sleeping_agent.restricted_roles = restricted_jobs
```
Comment (AnturK) [inline code/game/gamemodes/overthrow/overthrow.dm]: You're not setting special role anywhere so it's possible for them to be kicked back to lobby and other unpleasantness.

File: code/modules/antagonists/overthrow/overthrow.dm [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+	var/static/list/possible_useful_items
+
+/datum/antagonist/overthrow/New()
+	..()
+	if(!possible_useful_items)
+		possible_useful_items = list(/obj/item/gun/ballistic/automatic/pistol, /obj/item/storage/box/syndie_kit/throwing_weapons, /obj/item/pen/edagger, /obj/item/pen/sleepy, \
+									/obj/item/soap/syndie, /obj/item/card/id/syndicate, /obj/item/storage/box/syndie_kit/chameleon)
+
+/datum/antagonist/overthrow/on_gain()
+	create_team()
```
Comment (AnturK) [inline code/modules/antagonists/overthrow/overthrow.dm]: You don't need to call it manually, `create_team` is called in `add_antag_datum`, you should be passing the already existing team there. Personally i feel like splitting naming and team creation is good idea here.

File: code/modules/antagonists/overthrow/overthrow.dm [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+			traitor_mob.dna.add_mutation(CLOWNMUT)
+	..()
+// there is no promote command because there's no difference between a boss and the normal dude other than roundstart getting the items
+/datum/antagonist/overthrow/get_admin_commands()
+	. = ..()
+	.["Give random item"] = CALLBACK(src,.proc/equip_overthrow)
+	.["Give overthrow boss equip"] = CALLBACK(src,.proc/equip_overthrow_boss)
+
+/datum/antagonist/overthrow/proc/update_overthrow_icons_added(datum/mind/traitor_mind)
+	var/datum/atom_hud/antag/overthrowhud = GLOB.huds[team.hud_entry_num]
```
Comment (AnturK) [inline code/modules/antagonists/overthrow/overthrow.dm]: You should probably remove these at some point if team gets deleted.

File: code/modules/antagonists/overthrow/overthrow.dm [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+		to_chat(H, "Use the implanter in your [where] to wake up sleeping syndicate agents, so that they can aid you.")
+
+/datum/antagonist/overthrow/get_team()
+	return team
+
+/datum/antagonist/overthrow/greet()
+	to_chat(owner.current, "<B><font size=3 color=red>You are a syndicate sleeping agent. Your job is to stage a swift, fairly bloodless coup. </font></B>")
+
+/datum/antagonist/overthrow/boss
+	name = "Syndicate initial eutineers"
```
Comment (AnturK) [inline code/modules/antagonists/overthrow/overthrow.dm]: typo

File: code/modules/antagonists/overthrow/overthrow_team.dm [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+	var/hud_entry_num // A number holding the hud's index inside 'huds' global list. Gets set on hud update, if a hud doesn't exist already. Must be a number, otherwise BYOND shits up with assoc lists and everything goes to hell.
+
+/datum/team/overthrow/proc/create_objectives()
+	// Heads objective
+	var/datum/objective/ovethrow_heads/heads = new()
+	heads.team = src
+	heads.find_target()
+	objectives += heads
+	// AI objective
+	var/datum/objective/overthrow_AI/AI = new()
```
Comment (AnturK) [inline code/modules/antagonists/overthrow/overthrow_team.dm]: Didn't set owner here.

File: code/game/gamemodes/overthrow/objective.dm [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+/datum/objective/ovethrow_heads/update_explanation_text()
+	explanation_text = "Convert, exile or kill "
+	explanation_text += targets.Join(",")
+	explanation_text += ". Converting to your team will give you more points, whereas killing will give you the least. Syndicates don't want to stir up too many troubles."
+
+/datum/objective/ovethrow_heads/check_completion() // This won't return a boolean as usual, but the amount of points earned.
+	var/base_points = 0
+	for(var/i in targets)
+		var/datum/mind/M = i
+		if(M && M.current)
```
Comment (AnturK) [inline code/game/gamemodes/overthrow/objective.dm]: Blown up don't count ?

**Review by AnturK — CHANGES_REQUESTED** [review]

**Comment by drderp3635** [comment]
there are two suggestions i have for this.
1-give some kind of point value to regular crew members, a half point would probably do
2-give a war-type option to the teams, the leaders would have some form of voting for this, and if all leaders unanamously agree to it, it will make an anouncement to the station and the teams will have to surpass some number of points which requires nearly everyone to be converted.

File: code/game/gamemodes/overthrow/overthrow.dm (hunk 1 above)
Comment (Carbonhell) [reply code/game/gamemodes/overthrow/overthrow.dm]: that's a thing?what the fuck, why does the special_role var still exist
anyway,fixing

**Comment by Carbonhell** [comment]
Changes suggested by Antur applied, but i still need to properly test this.

File: code/game/gamemodes/overthrow/overthrow.dm [hunk 7]
```diff
@@ -0,0 +1,47 @@
+/datum/game_mode/overthrow
+	name = "overthrow"
+	config_tag = "overthrow"
+	antag_flag = ROLE_OVERTHROW
+	restricted_jobs = list("Security Officer", "Warden", "Detective", "AI", "Cyborg","Captain", "Head of Personnel", "Head of Security", "Chief Engineer", "Research Director", "Chief Medical Officer")
+	required_players = 1 //20 // the core idea is of a swift, bloodless coup, so it shouldn't be as chaotic as revs.
```
Comment (Carbonhell) [inline code/game/gamemodes/overthrow/overthrow.dm]: gotta reset those values once the PR is ready

**Comment by Carbonhell** [comment]
@drderp3635 The syndies already have a benefit from recruiting random people,aka having more people to do their objectives with. I don't think points are necessary, plus you can only convert up to 2-3 guys and you shouldn't be penalized if you cannot find someone to convert imo
the wartype option is completely outside the gamemode's idea of being a bloodless silent coup, i don't think it fits at all honestly

**Comment by Carbonhell** [comment]
okay i don't know what the hell happened here

**Comment by Carbonhell** [comment]
nvm PR is still broken due to 3 unrelated files slipping in,welp

**Comment by Carbonhell** [comment]
I'm gonna reopen this because i have no idea what i did with git/how to fix this

**Pull request closed without merging**